### ISSUE TITLE
PR 892 does not always work

### ISSUE THREAD

**Issue by mvorisek** [sebastianbergmann/phpcov#122]
| Q                         | A
| --------------------------| ---------------
| php-code-coverage version | 9.2.17 (latest)
| PHP version               | 8.0.23 (latest)
| Driver                    | Xdebug
| Xdebug version  | 3.2.0alpha3 (latest)
| Installation Method       | Composer
| Usage Method              | PHPUnit & other
| PHPUnit version (if used) | 9.5.24 (latest)

An issue, that took me quite a while to figure out what is going on.

In atk4/ui we have the following coverage setup:

a) collect coverage in phpunit tests, config: https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/phpunit.xml.dist#L23-L29
b) collect coverage from web requests, config: https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/demos/init-app.php#L21-L24 and https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/src/Behat/CoverageUtil.php#L38-L43
c) then merge all the coverage files using phpunit/phpcov in https://github.com/atk4/ui/blob/33779a9414b624a8faa7e959883d4ba407099dcd/.github/workflows/test-unit.yml#L223

Coverage collection is working. But as soon as we include `tests/` dir also in b), the coverage decreases. This is wrong, as all `tests/` files are already included in a) and inclusion of an already included file in another coverage collection cannot decrease the merged coverage.

CI repro: https://github.com/atk4/ui/pull/1852/commits - see 2nd and 3rd commit

https://app.codecov.io/gh/atk4/ui/blob/b762d62434df393aad170a7378d3d45876f93f5a/tests/JsTest.php (2nd commit coverage)
https://app.codecov.io/gh/atk4/ui/blob/99f347b7d04567faefafa3b39105042b728e1649/tests/JsTest.php (3rd commit coverage)

It seems the problem is in https://github.com/sebastianbergmann/php-code-coverage/pull/892 or in https://github.com/sebastianbergmann/phpcov.

**Comment by Slamdunk** [comment]
The code provided produces two identical files, both with `<line num="32" type="stmt" count="0"/>`

**Comment by mvorisek** [comment]
> The code provided produces two identical files, both with `<line num="32" type="stmt" count="0"/>`

@Slamdunk can you please double check?

If l32 is present in `out.xml` as well, this will imply https://github.com/sebastianbergmann/php-code-coverage/pull/892 not applied at all. Do you use the latest deps?

Here are attach the output files from my end:

[out.xml](https://github.com/sebastianbergmann/phpcov/files/9565585/out.xml.txt)
[outM.xml](https://github.com/sebastianbergmann/phpcov/files/9565586/outM.xml.txt)

**Comment by Slamdunk** [comment]
My best ipothesis is that the well known hack of PHPUnit to use `include`/`require` to parse unknown classes is messing things up here.

Please try these two changes together:

1. Add [`processUncoveredFiles`](https://github.com/sebastianbergmann/phpunit/blob/d0aa6097bef9fd42458a9b3c49da32c6ce6129c5/phpunit.xsd#L35) to your `phpunit.xml`
2. Add `self::$coverage->processUncoveredFiles();` to your `Atk4\Ui\Behat\CoverageUtil::start()`

And let's see how the CI behaves.

**Comment by mvorisek** [comment]
@Slamdunk please see my https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1244616184 post here - in a few lines, I have provided a script to reproduce - all you need is `phpunit/php-code-coverage` and https://github.com/atk4/ui/blob/01d39d24694e4b4f63a576d090e55e0e97aaa9e4/tests/JsTest.php file, no phpunit even needed

it should output simillar files as attached in https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1246580539 with one correct and one without https://github.com/sebastianbergmann/php-code-coverage/pull/892 applied - that is a bug that needs to be fixed.

**Comment by Slamdunk** [comment]
That's indeed what I did: I'm now waiting to see if what's I suggested resolves your issue in `atk4/ui`

**Comment by mvorisek** [comment]
@Slamdunk according https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1246288809 you claims you are not able to reproduce it.

To prevent any misunderstandings, I retried on a clean directory with the following steps:

1. `composer require phpunit/php-code-coverage`
2. save https://github.com/atk4/ui/blob/01d39d24694e4b4f63a576d090e55e0e97aaa9e4/tests/JsTest.php inside the current directory and delete ` extends TestCase` from the file (otherwise file cannot be included when the base class is not available)
3. save:
   ```
   <?php
   
   declare(strict_types=1);
   
   namespace Atk4\Ui\ReproCoverage;
   
   use SebastianBergmann\CodeCoverage\CodeCoverage;
   use SebastianBergmann\CodeCoverage\Driver\Selector as DriverSelector;
   use SebastianBergmann\CodeCoverage\Filter;
   use SebastianBergmann\CodeCoverage\Report;
   
   require_once __DIR__ . '/vendor/autoload.php';
   
   $filePath = __DIR__ . '/JsTest.php';
   
   $filter = new Filter();
   $filter->includeFile($filePath);
   
   $cov = new CodeCoverage((new DriverSelector())->forLineCoverage($filter), $filter);
   $cov->start('xxx');
   
   // ⚠️ no include implies https://github.com/sebastianbergmann/php-code-coverage/pull/892 not applied
   require_once $filePath;
   
   $cov->stop();
   
   $covM = new CodeCoverage((new DriverSelector())->forLineCoverage($filter), $filter);
   $writer = new Report\Clover();
   $writer->process($cov, __DIR__ . '/out.xml');
   $writer->process($covM, __DIR__ . '/outM.xml');
   ```
   as `repro.php`
4. run `phpw . -d xdebug.mode=coverage repro.php` (make sure xdebug is enabled and the script should produce empty result without any error)

![image](https://user-images.githubusercontent.com/2228672/190381246-9cae6520-fde1-498b-a20e-320e5b18855e.png)

**Comment by Slamdunk** [comment]
> I retried on a clean directory with the following steps

Nice, and have you also tried the `processUncoveredFiles` directive as indicated in the docs https://phpunit.readthedocs.io/en/9.5/code-coverage-analysis.html#including-files
?

**Comment by mvorisek** [comment]
@Slamdunk please rerun the repro code using the exact steps I provided, the repro code contains all the settings in the code - you should end up with two xml files, one with l32 coverage file, one without.

If you still see l32 in both files (I very doubt so), then please tell me how to adjust this specific repro.php file so https://github.com/sebastianbergmann/php-code-coverage/pull/892 is applied at least in one case.

**Comment by Slamdunk** [comment]
Simply call `processUncoveredFiles` to both `$cov` and `$covM`

**Comment by mvorisek** [comment]
with `$cov->processUncoveredFiles()` (placed after `new CodeCoverage(...` and for `$covM` too) I get `outM.xml` /wo `JsTest.php` at all

[outM.xml.txt](https://github.com/sebastianbergmann/phpcov/files/9575283/outM.xml.txt)

I would still like if you can run the repro code using the steps from https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1247907753. The discussion will be much easier if you have the problem on your side too.

**Comment by Slamdunk** [comment]
The only important thing here is that line 32 must not appear in neither `out.xml` nor `outM.xml`: `processUncoveredFiles` achieve exactly that

**Comment by mvorisek** [comment]
@Slamdunk no, we need the file to be present/analysed to provide informations like uncovered methods etc.

With https://github.com/sebastianbergmann/php-code-coverage/pull/892 I would say l32 should not be present in any output.

Please reproduce using the steps provided in https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1247907753.

**Comment by Slamdunk** [comment]
I'm sorry I think I'm not capable of help you futher

**Comment by mvorisek** [comment]
@Slamdunk this discussion here is because of your https://github.com/sebastianbergmann/php-code-coverage/pull/892 PR

`processUncoveredFiles` implies all files are included which implies some files can be executed, let's please do not discuss this option

I point out a problem that in some cases https://github.com/sebastianbergmann/php-code-coverage/pull/892 PR applies and in some not. Is this expected? I belive not.

In https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1247907753 I have provided a repro case. Would you be please so kind and run it on your side to confirm the problem?

**Comment by mvorisek** [comment]
> @mvorisek btw, as soon as you are able to give us a reproducible local environment for the issue, I can debug it quickly. But without such premise, I can't help further


https://github.com/sebastianbergmann/phpcov/issues/122#issuecomment-1246288809 I am still convinced you must have been doing something wrong or using pre-892 PR deps

@Slamdunk anyway, since the issue is clear now, I am able to propose a PR with a fix. With 892 PR, your intentions were it will apply to all output data, right?

**Comment by mvorisek** [comment]
@sebastianbergmann please move to https://github.com/sebastianbergmann/php-code-coverage, it needs to be fixed there
